hw3/hw3.cpp

```diff
@@ -0,0 +1,214 @@
+            pixelValue.x += (float)(in[(row+u) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+            pixelValue.y += (float)(in[((row+u) + height) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+            pixelValue.z += (float)(in[((row+u) + height*2) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+         }
+       }
+
+   }
+    
+   //Load rgb values into output[row][col]
+   out[row * width + col] = pixelValue.x > 255 ? 255 : pixelValue.x;
+   out[(row + height) * width + col] = pixelValue.y > 255 ? 255 : pixelValue.y;
+   out[(row + height*2) * width + col] = pixelValue.z > 255 ? 255 : pixelValue.z;
+}
+
+
+__global__ void L2Kernel(const unsigned char *blurImg, int width, int height, const unsigned char *origImg, GMat L2)
+{
+    int row = blockIdx.y * blockDim.y + threadIdx.y;
+    int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float3 pixelValueB;
+    pixelValueB.x = 0.0f;
+    pixelValueB.y = 0.0f;
+    pixelValueB.z = 0.0f;
+    pixelValueB.x = (float)(blurImg[row * width + col]);
+    pixelValueB.y = (float)(blurImg[(row + height) * width + col]);
+    pixelValueB.z = (float)(blurImg[(row + height*2) * width + col]);
+
+    float3 pixelValueO;
+    pixelValueO.x = 0.0f;
+    pixelValueO.y = 0.0f;
+    pixelValueO.z = 0.0f;
+    pixelValueO.x = (float)(origImg[row * width + col]);
+    pixelValueO.y = (float)(origImg[(row + height) * width + col]);
+    pixelValueO.z = (float)(origImg[(row + height*2) * width + col]);
+
+    float temp1 = abs(pixelValueB.x - pixelValueO.x);
+    temp1 *= temp1;
+    float temp2 = abs(pixelValueB.y - pixelValueO.y);
+    temp2 *= temp2;
+    float temp3 = abs(pixelValueB.z - pixelValueO.z);
+    temp3 *= temp3;
+    temp1 += temp2 + temp3;
+    temp1 = temp1/3.f;
+
+    L2.elements[row*L2.width+col] = temp1;
+
+}
+
+
+__global__ void reduce(GMat dOut, GMat dIn)
+{
+    extern __shared__ float sdata[];
+    int myId = threadIdx.x + blockDim.x*blockIdx.x;
+    int tid  = threadIdx.x;
+    sdata[tid] = dIn.elements[myId];
+    __syncthreads();
+    for(int s = 1; s<blockDim.x; s*=2)
+    {
+	if(tid%(2*s)==0)
+        {
+	   sdata[tid] += sdata[tid+s];
+        }
+	__syncthreads();
+    }
+    if(tid == 0)
+    {
+        dOut.elements[blockIdx.x] = sdata[0];
+    }
+
+
+}
+
+
+void L2Norm(const unsigned char *blurImg, int width, int height, const unsigned char *origImg)
+{
+   CMat CL2(width, height);
+   GMat L2(width, height);
+   dim3 dimBlock(BLOCK_SIZE, BLOCK_SIZE);
+   dim3 dimGrid(width / dimBlock.x+1, height / dimBlock.y);
+   L2Kernel<<<dimGrid, dimBlock>>>(blurImg,width,height,origImg,L2);
+   CL2.load(L2);
+   //Reduce
+   CMat Creduced(1024,1);
+   GMat reduced(1024,1);
+   dim3 dimBlockReduce(1024);
+   dim3 dimGridReduce((width*height+dimBlockReduce.x-1)/dimBlockReduce.x);
+   reduce<<<dimGridReduce,dimBlockReduce,1024*sizeof(float)>>>(reduced,L2);
+   reduce<<<1,dimBlockReduce,1024*sizeof(float)>>>(reduced,reduced);
```

Comment:
We'll need more stages of reduce here. width*height > 1024^2. In fact, the image is 1199x1200 I'm also not seeing a sqrt here.